docs/pipeline-components-and-applications/loaders-storage-targets/bigquery-loader/upgrade-guides/2-0-0-upgrade-guide/index.md

```diff
@@ -0,0 +1,92 @@
+
+:::tip Consolidating old and new columns
+
+If you are using [Snowplow dbt models](/docs/modeling-your-data/modeling-your-data-with-dbt/dbt-models/index.md), they will automatically consolidate the data between `_1_0_0` and `_1` style columns, because they look at the major version prefix (e.g. `_1`), which is common to both.
+
+If you are not using Snowplow dbt models but still use dbt, you can employ [this macro](https://github.com/snowplow/dbt-snowplow-utils#combine_column_versions-source) to manually aggregate the data across old and new columns.
+
+:::
+
+## Recovery columns
+
+### What is schema evolution?
+
+One of Snowplow’s key features is the ability to [define custom schemas and validate events against them](https://docs.snowplow.io/docs/understanding-tracking-design/understanding-schemas-and-validation/). Over time, users often evolve the schemas, e.g. by adding new fields or changing existing fields. To accommodate these changes, BigQuery Loader 2.0.0 automatically adjusts the database tables in the warehouse accordingly.
+
+There are two main types of schema changes:
+
+**Breaking**: The schema version has to be changed in a major way (`1-2-3` → `2-0-0`). As of BigQuery Loader 2.0.0, each major schema version has its own column (`..._1`, `..._2`, etc, for example: `contexts_com.snowplowanalytics_ad_click_1`).
+
+**Non-breaking**: The schema version can be changed in a minor way (`1-2-3` → `1-3-0` or `1-2-3` → `1-2-4`). Data is stored in the same database column.
+
+Loader tries to format the incoming data according to the latest version of the schema it saw (for a given major version, e.g. `1-*-*`). For example, if a batch contains events with schema versions `1-0-0`, `1-0-1` and `1-0-2`, the loader derives the output schema based on version `1-0-2`. Then the loader instructs BigQuery to adjust the database column and load the data.
+
+### Recovering from invalid schema evolution 
+
+Let's consider these two schemas as an example of breaking schema evolution (changing the type of a field from `integer` to `string`) using the same major version (`1-0-0` and `1-0-1`):
+
+```json
+{
+   // 1-0-0
+   "properties": {
+      "a": {"type": "integer"}
+   }
+}
+```
+
+```json
+{
+   // 1-0-1
+   "properties": {
+      "a": {"type": "string"}
+   }
+}
+```
+
+Because `1-0-1` events cannot be loaded into the same column with `1-0-0`, the data would be put in a separate column, e.g. `contexts_com_snowplowanalytics_ad_click_1_0_1_recovered_9999999`, where:
+  - `1_0_1` is the version of the offending schema;
+  - `9999999` is a hash code unique to the schema (i.e. it will change if the schema is overwritten with a different one).
+
+If you create a new schema `1-0-2` that reverts the offending changes and is again compatible with `1-0-0`, the data for events with that schema will be written to the original column as expected.
+
```

Comment:
```suggestion
:::tip
You might find that some of your schemas were evolved incorrectly in the past, which results in the creation of these “recovery” columns after the upgrade. To address this for a given schema, create a new _minor_ schema version that reverts the breaking changes introduced in previous versions. (Or, if you want to keep the breaking change, create a new _major_ schema version.) You can set it to [supersede](/docs/understanding-tracking-design/versioning-your-data-structures/amending/index.md#marking-the-schema-as-superseded) the previous version(s), so that events are automatically validated against the new schema.
:::
```

Do we also need instructions similar to these? https://docs.snowplow.io/docs/pipeline-components-and-applications/loaders-storage-targets/snowplow-rdb-loader/upgrade-guides/6-0-0-upgrade-guide/#identifying-schemas-that-need-patching @istreeter

Reply:
That section "Identifying schemas that need patching" was written long ago before we had the supercededBy feature.  I would think now we have supercededBy, we shouldn't recommend to anyone to patch schemas.

Instead, the user can publish the fixed schema as a new major version, and configure the old schemas to automatically upgrade to the new major version.  This is consistent with our belief that schemas are immutable.  And it means we don't need to write a new igluctl feature.

Reply:
My doubt was not specifically about patching but whether running igluctl would be useful to identify issues with schemas... Because I think it says what the incompatibilities are? With just a recovery column, how will the users know what to fix?

Reply:
igluctl has an existing sub-command `verify-parquet`

```
igluctl static verify-parquet <schema_dir>
```

It prints out warnings like

```
'com.test/test/jsonschema/1-0-1'. Changes: Incompatible type change String to Double at /id"
```

Unfortunately, it only works if the schemas are written into files on the local file system.  It does not connect to remote Iglus.

We've not really planned what happens when existing users migrate to the new loader.  And that is semi-deliberate: it is a lot of work to gather all the requirements and plan for all the things that could go wrong.  So I've been delaying it until there is time to work on it properly.

Reply:
@stanch @istreeter as BQ Loader v2 relies on the same schema compatibility model/rules as parquet/Databricks/lake loader, we could use `verify-parquet` igluctl command and recommend it to users to detect incompatibilities. But this `parquet` in the name is confusing! Should we also do some renaming in igluctl then to make it more generic? And then also adjust [docs](https://deploy-preview-877--snowplow-docs.netlify.app/docs/pipeline-components-and-applications/iglu/igluctl-2/#igluctl)